HTTPS does not work through Proxy

I had originally reported this against npm, but it looks like the issue is in request. Here's the npm ticket: https://github.com/isaacs/npm/issues/2866

I have set up my `.npmrc` file to have settings for our corporate proxy:

```
proxy = http://user:password@proxy:8080/
https-proxy = http://user:password@proxy:8080/
```

Running any npm command using these settings fails. As a workaround, I had to change the registry to HTTP instead of HTTPS:

```
registry = http://registry.npmjs.org/
```

Most things seem to work fine with this, except if I try to install something that has dependencies listed using HTTPS. One example for this is Yeoman (https://github.com/yeoman/yeoman), which lists two dependencies in package.json using HTTPS:

```
https://nodeload.github.com/cowboy/grunt/tarball/0ba6d4b529
https://nodeload.github.com/yeoman/generators/tarball/282b0b4c51
```

Downloading these fails miserably with the following error:

```
npm ERR! fetch failed https://nodeload.github.com/cowboy/grunt/tarball/0ba6d4b529
npm ERR! fetch failed https://nodeload.github.com/yeoman/generators/tarball/282b0b4c51
npm ERR! Error: tunneling socket could not be established, sutatusCode=403
npm ERR!     at ClientRequest.onConnect (/usr/local/lib/node_modules/npm/node_modules/request/tunnel.js:148:19)
npm ERR!     at ClientRequest.g (events.js:193:14)
npm ERR!     at ClientRequest.EventEmitter.emit (events.js:123:20)
npm ERR!     at Socket.socketOnData (http.js:1393:11)
npm ERR!     at TCP.onread (net.js:403:27)
npm ERR! If you need help, you may report this log at:
npm ERR!     <http://github.com/isaacs/npm/issues>
npm ERR! or email it to:
npm ERR!     <npm-@googlegroups.com>

npm ERR! System Darwin 10.8.0
npm ERR! command "node" "/usr/local/bin/npm" "install" "yeoman"
npm ERR! cwd /Users/Nils.Winkler
npm ERR! node -v v0.8.11
npm ERR! npm -v 1.1.62
npm ERR! code ECONNRESET
```

It seems like HTTPS does not work through some proxies. Could you please look into getting this supported in npm, as it looks like a major shortcoming for people behind a corporate proxy.

--- a/main.js
+++ b/main.js
@@ -138,6 +138,7 @@ Request.prototype.init = function (options) {
   } else {
     if (typeof self.uri == "string") self.uri = url.parse(self.uri)
   }
+  
   if (self.proxy) {
     if (typeof self.proxy == 'string') self.proxy = url.parse(self.proxy)
 
@@ -148,7 +149,9 @@ Request.prototype.init = function (options) {
 
       var tunnelOptions = { proxy: { host: self.proxy.hostname
                                    , port: +self.proxy.port
-                                   , proxyAuth: self.proxy.auth }
+                                   , proxyAuth: self.proxy.auth
+                                   , headers: { Host: self.uri.hostname + ':' + 
+                                        (self.uri.port || self.uri.protocol === 'https:' ? 443 : 80) }}
                           , ca: this.ca }
 
       self.agent = tunnelFn(tunnelOptions)

--- a/tests/test-tunnel.js
+++ b/tests/test-tunnel.js
@@ -23,7 +23,7 @@ var ready = false
 squid.stderr.on('data', function (c) {
   console.error('SQUIDERR ' + c.toString().trim().split('\n')
                .join('\nSQUIDERR '))
-  ready = c.toString().match(/ready to serve requests/i)
+  ready = c.toString().match(/ready to serve requests|Accepting HTTP Socket connections/i)
 })
 
 squid.stdout.on('data', function (c) {

